File: src/server/services/agent-context.ts
import { resolve } from "node:path";
import type {
	AgentContextLayer,
	AgentContextMode,
	AgentContextUpdate,
	EffectiveAgentContext,
} from "../../features/agent-context/types.ts";
import type { Prompt } from "../../features/prompts/types.ts";
import { atomicWriteJson, readJson } from "../../lib/route-helpers.ts";
import { PROJECT_ROOT, userDataPath } from "../../lib/user-data.ts";

const CONTEXT_FILE = userDataPath("agent-context.json");
const LOCAL_SKILLS_FILE = userDataPath("prompts.json");
const BUNDLED_SKILLS_FILE = resolve(PROJECT_ROOT, "data/prompts.json");

interface AgentContextFile {
	global: AgentContextLayer;
	projects: Record<string, AgentContextLayer>;
	chats: Record<string, AgentContextLayer>;
}

const EMPTY_LAYER: AgentContextLayer = {
	instructions: "",
	mode: "inherit",
	updatedAt: 0,
};

function normalizeMode(value: unknown): AgentContextMode {
	return value === "replace" ? "replace" : "inherit";
}

function normalizeLayer(value: unknown): AgentContextLayer {
	if (!value || typeof value !== "object") return { ...EMPTY_LAYER };
	const layer = value as Partial<AgentContextLayer>;
	return {
		instructions:
			typeof layer.instructions === "string" ? layer.instructions : "",
		mode: normalizeMode(layer.mode),
		updatedAt: typeof layer.updatedAt === "number" ? layer.updatedAt : 0,
	};
}

async function loadContextFile(): Promise<AgentContextFile> {
	const stored = await readJson<Partial<AgentContextFile>>(CONTEXT_FILE, {});
	return {
		global: normalizeLayer(stored.global),
		projects:
			stored.projects && typeof stored.projects === "object"
				? Object.fromEntries(
						Object.entries(stored.projects).map(([key, value]) => [
							key,
							normalizeLayer(value),
						])
					)
				: {},
		chats:
			stored.chats && typeof stored.chats === "object"
				? Object.fromEntries(
						Object.entries(stored.chats).map(([key, value]) => [
							key,
							normalizeLayer(value),
						])
					)
				: {},
	};
}

function projectKey(cwd?: string): string | null {
	if (!cwd?.trim()) return null;
	return resolve(cwd.trim());
}

function composeLayers(
	global: AgentContextLayer,
	project: AgentContextLayer | null,
	chat: AgentContextLayer | null
): string {
	let parts = global.instructions.trim() ? [global.instructions.trim()] : [];
	for (const layer of [project, chat]) {
		if (!layer?.instructions.trim()) continue;
		parts =
			layer.mode === "replace"
				? [layer.instructions.trim()]
				: [...parts, layer.instructions.trim()];
	}
	return parts.join("\n\n");
}

export function createSkillManifest(skills: readonly Prompt[]): string {
	if (skills.length === 0) return "";
	return skills
		.map(
			(skill) =>
				`- ${skill.command}: ${skill.description || skill.name} (invoke with /${skill.command} or $${skill.command})`
		)
		.join("\n");
}

function activatedSkills(skills: readonly Prompt[], text: string) {
	const normalized = text.toLowerCase();
	return skills.flatMap((skill) => {
		const explicit =
			normalized.includes(`/${skill.command.toLowerCase()}`) ||
			normalized.includes(`$${skill.command.toLowerCase()}`);
		const triggerTerms = [skill.command.replaceAll("-", " "), skill.name]
			.map((value) => value.toLowerCase().trim())
			.filter((value) => value.length >= 4);
		const automatic = triggerTerms.some((term) => normalized.includes(term));
		return explicit || automatic
			? [
					{
						name: skill.name,
						command: skill.command,
						instructions: skill.promptTemplate,
					},
				]
			: [];
	});
}

async function loadSkillsForManifest(): Promise<Prompt[]> {
	const sources = await Promise.all(
		[BUNDLED_SKILLS_FILE, LOCAL_SKILLS_FILE].map(async (path) => {
			const file = Bun.file(path);
			if (!(await file.exists())) return [];
			try {
				return JSON.parse(await file.text()) as Prompt[];
			} catch {
				return [];
			}
		})
	);
	const byCommand = new Map<string, Prompt>();
	for (const skill of sources.flat()) byCommand.set(skill.command, skill);
	return [...byCommand.values()];
}

let writeQueue: Promise<unknown> = Promise.resolve();

export const AgentContextService = {
	async resolveForAgent(cwd?: string, paneId?: string, text = "") {
		const skills = await loadSkillsForManifest();
		const context = await this.resolve(cwd, paneId, skills);
		return { ...context, activatedSkills: activatedSkills(skills, text) };
	},
	async resolve(
		cwd: string | undefined,
		paneId: string | undefined,
		skills: readonly Prompt[] = []
	): Promise<EffectiveAgentContext> {
		const stored = await loadContextFile();
		const key = projectKey(cwd);
		const project = key ? (stored.projects[key] ?? null) : null;
		const chat = paneId ? (stored.chats[paneId] ?? null) : null;
		return {
			global: stored.global,
			project,
			chat,
			effectiveInstructions: composeLayers(stored.global, project, chat),
			scope: chat?.instructions.trim()
				? "chat"
				: project?.instructions.trim()
					? "project"
					: "global",
			skillCount: skills.length,
			skillManifest: createSkillManifest(skills),
			activatedSkills: [],
		};
	},

	async update(update: AgentContextUpdate): Promise<void> {
		const operation = writeQueue.then(async () => {
			const stored = await loadContextFile();
			const layer: AgentContextLayer = {
				instructions: update.instructions.trim(),
				mode: normalizeMode(update.mode),
				updatedAt: Date.now(),
			};
			if (update.scope === "global") stored.global = layer;
			else if (update.scope === "project") {
				const key = projectKey(update.cwd);
				if (!key) throw new Error("A project directory is required");
				if (layer.instructions) stored.projects[key] = layer;
				else delete stored.projects[key];
			} else {
				if (!update.paneId) throw new Error("A chat pane is required");
				if (layer.instructions) stored.chats[update.paneId] = layer;
				else delete stored.chats[update.paneId];
			}
			await atomicWriteJson(CONTEXT_FILE, stored, 2);
		});
		writeQueue = operation.catch(() => {});
		await operation;
	},
};

File: native/core/src/agent_context.rs
//! Context composition and skill activation. Project keys are normalized by the caller.
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::prompts::Prompt;

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct AgentContextLayer {
    pub instructions: String,
    #[ts(type = "'inherit' | 'replace'")]
    pub mode: String,
    pub updated_at: u64,
}

#[derive(Debug, Serialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct EffectiveAgentContext {
    pub global: AgentContextLayer,
    pub project: Option<AgentContextLayer>,
    pub chat: Option<AgentContextLayer>,
    pub effective_instructions: String,
    #[serde(skip)]
    pub activated_skills: String,
}

#[derive(Debug, ts_rs::TS)]
#[ts(rename_all = "camelCase")]
pub struct AgentContextUpdate {
    #[ts(type = "'global' | 'project' | 'chat'")]
    pub scope: String,
    #[ts(optional)]
    pub cwd: Option<String>,
    #[ts(optional)]
    pub pane_id: Option<String>,
    pub instructions: String,
    #[ts(optional, type = "'inherit' | 'replace'")]
    pub mode: Option<String>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct AgentContextState {
    global: AgentContextLayer,
    projects: BTreeMap<String, AgentContextLayer>,
    chats: BTreeMap<String, AgentContextLayer>,
}

impl AgentContextState {
    pub fn resolve(
        mut self,
        project_key: Option<&str>,
        pane_id: Option<&str>,
    ) -> EffectiveAgentContext {
        let project = project_key.and_then(|key| self.projects.remove(key));
        let chat = pane_id.and_then(|id| self.chats.remove(id));
        let effective_instructions = compose_layers(&self.global, project.as_ref(), chat.as_ref());
        EffectiveAgentContext {
            global: self.global,
            project,
            chat,
            effective_instructions,
            activated_skills: String::new(),
        }
    }

    pub fn resolve_for_agent(
        self,
        project_key: Option<&str>,
        pane_id: Option<&str>,
        text: &str,
        skills: &[Prompt],
    ) -> EffectiveAgentContext {
        let mut context = self.resolve(project_key, pane_id);
        let normalized = text.to_lowercase();
        context.activated_skills = skills
            .iter()
            .filter(|skill| {
                normalized.contains(&format!("/{}", skill.command.to_lowercase()))
                    || [skill.command.replace('-', " ").as_str(), &skill.name]
                        .into_iter()
                        .any(|term| {
                            let lower = term.to_lowercase();
                            let term = lower.trim();
                            term.len() >= 4 && normalized.contains(term)
                        })
            })
            .map(|skill| {
                format!(
                    "<activated-skill name=\"{}\">\n{}\n</activated-skill>",
                    skill.command, skill.prompt_template
                )
            })
            .collect::<Vec<_>>()
            .join("\n\n");
        context
    }

    pub fn update(&mut self, update: AgentContextUpdate, now: u64) -> Result<(), String> {
        let layer = AgentContextLayer {
            instructions: update.instructions.trim().to_string(),
            mode: normalize_mode(update.mode.as_deref()),
            updated_at: now,
        };
        match update.scope.as_str() {
            "global" => self.global = layer,
            "project" | "chat" => {
                let (layers, key) = if update.scope == "project" {
                    (
                        &mut self.projects,
                        update
                            .cwd
                            .filter(|key| !key.is_empty())
                            .ok_or_else(|| "A project directory is required".to_string())?,
                    )
                } else {
                    (
                        &mut self.chats,
                        update
                            .pane_id
                            .filter(|value| !value.is_empty())
                            .ok_or_else(|| "A chat pane is required".to_string())?,
                    )
                };
                if layer.instructions.is_empty() {
                    layers.remove(&key);
                } else {
                    layers.insert(key, layer);
                }
            }
            _ => return Err("scope is invalid".into()),
        }
        Ok(())
    }
}

impl Default for AgentContextLayer {
    fn default() -> Self {
        Self {
            instructions: String::new(),
            mode: "inherit".into(),
            updated_at: 0,
        }
    }
}

fn normalize_mode(value: Option<&str>) -> String {
    if value == Some("replace") {
        "replace".into()
    } else {
        "inherit".into()
    }
}

fn compose_layers(
    global: &AgentContextLayer,
    project: Option<&AgentContextLayer>,
    chat: Option<&AgentContextLayer>,
) -> String {
    let mut parts = Vec::new();
    for layer in [Some(global), project, chat].into_iter().flatten() {
        let instructions = layer.instructions.trim();
        if instructions.is_empty() {
            continue;
        }
        if layer.mode == "replace" {
            parts.clear();
        }
        parts.push(instructions);
    }
    parts.join("\n\n")
}
